[Azure Search] Upgrade data plane to latest AutoRest and split into 4 packages

bitsforthought opened the pull request:
This PR includes some big changes to the Azure Search data plane SDK:
  - Upgraded to the latest version of AutoRest.
  - Split into four NuGet packages so customers can use different subsets of the functionality as needed.
  - Bumped major version to 5.0-preview
  - Fixed a new compatibility issue with JSON.NET 11 that was recently reported.

See individual commits for details.

Note that this new version is not ready to ship yet. We still have more features to add before we will publish these new packages, which we will do out of our own preview branch.

FYI @mhko @chaosrealm @natinimni 

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers that changes made by the pull request.

If you are regenerating your SDK based off of a new swagger spec, please add the link to the corresponding swagger spec pull request that has been merged in the azure-rest-api-specs repository
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

bitsforthought committed:
[Azure Search] Upgrade data plane to the latest version of AutoRest

bitsforthought committed:
[Azure Search] Splitting the data plane into four packages

What's motivating this change is a recognition that the data plane SDK is
really fulfilling three scenarios:

 1. Actually implementing search features in applications (full-text search,
    paging, faceted navigation, etc.)
 2. Indexing data via the push API
 3. Automating the management of indexes, synonyms, indexers, data sources,
    and other service-level resources.

Scenarios 1 and 2 are what you'd normally think of as "data plane". Scenario
3 is closer to "control plane" or management plane, or even "metadata plane"
since managing index schemas in particular is still pretty close to the data.

The split goes something like this:
 - Microsoft.Azure.Search.Data: Fulfills scenarios 1 and 2 above. This covers
   everything under the /docs resource. Currently most of this is custom code.
 - Microsoft.Azure.Search.Service: Fulfills scenario 3 above. This covers all
   top-level resources of the Azure Search Service REST API. This is mostly
   generated code, and our eventual goal is for nearly all of it to be
   generated.
 - Microsoft.Azure.Search.Common: Utilities common to Data and Service.
 - Microsoft.Azure.Search: A stub package that depends on all the others,
   plus defines the Indexes.GetClient() method that acts as a bridge from
   SearchServiceClient to SearchIndexClient. This package provides a
   convenient upgrade path for existing SDK users.

Eventually we're thinking of using Microsoft.Azure.Search.Service in
conjunction with Microsoft.Azure.Management.Search to implement a rich set
of PowerShell CmdLets that can manage indexes, data sources, indexers, etc.
in addition to the ARM-level resources.

bitsforthought committed:
[Azure Search] Factoring out data plane assembly and package versions

Also bumping all data plane package major versions since moving types to
different assemblies is a breaking change.

bitsforthought committed:
[Azure Search] Replacing ConverterBase with JsonReader extension methods

Reuse via inheritance is an anti-pattern, so we're getting rid of it.

bitsforthought committed:
[Azure Search] Fixing issue with copying JsonSerializerSettings

Since the Azure .NET SDKs have to support a wide range of JSON.NET versions,
it isn't safe to reference properties of JsonSerializerSettings by name since
these properties come and go between versions. Instead, we use reflection to
dynamically generate code to copy a JsonSerializerSettings instance.

brjohnstmsft commented:
One of the commits in this PR fixes the `MissingMethodException` issue mentioned here: https://github.com/Azure/azure-sdk-for-net/issues/3441

brjohnstmsft commented:
@isaacabraham This PR contains a fix for the `MissingMethodException` issue that you reported (among other things). I ended up solving it via reflection & dynamic code generation; Feel free to review if you're so inclined.

mhko commented on src/SDKs/Search/DataPlane/Microsoft.Azure.Search/GeneratedSearchIndex/Models/DocumentIndexResult.cs:
```diff
@@ -24,7 +24,10 @@ public partial class DocumentIndexResult
+        public DocumentIndexResult()
+        {
+            CustomInit();
+        }
```
Why do we have CustomInit(). Is there a model that defines it?

brjohnstmsft replied:
@mhko This is a new feature in AutoRest. `CustomInit` is a `partial` method, so if we want we can now write custom initialization code for generated model classes. This will be very useful in the future, but we aren't yet using this capability now.

Since `CustomInit` is not defined anywhere and is marked as `partial`, the compiler will omit the call.

mhko replied:
Thanks Bruce

---
In reply to: [172339646](https://github.com/Azure/azure-sdk-for-net/pull/4104#discussion_r172339646) [](ancestors = 172339646)

mhko commented:
LGTM

shahabhijeet commented on src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Microsoft.Azure.Search.Data.csproj:
```diff
@@ -0,0 +1,21 @@
+    <PackageReference Include="Microsoft.Spatial" Version="7.2.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.Azure.Search.Common\Microsoft.Azure.Search.Common.csproj" />
```
@brjohnstmsft the order in which you publish to nuget will matter even if you have project reference, just FYI

brjohnstmsft replied:
Good to know. I assume we'll need to publish dependencies first, then the packages that depend on them, correct?

shahabhijeet commented on src/SDKs/Search/DataPlane/AzSearch.DataPlane.props:
```diff
@@ -0,0 +1,10 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <TargetFrameworks>net452;netstandard1.4</TargetFrameworks>
```
@brjohnstmsft looks like the bug has been fixed that allows you specify Fx Version common for projects.
In that case, we will apply target Fx version from a common file and at that point, this line will be deleted from here (of course you will be involved in the PR), just as an FYI

brjohnstmsft replied:
Thanks!

shahabhijeet reviewed (changes requested):
With the new project structure, just an FYI, we will now not be able to run any tests during signing/publishing.
Because our simple heuristics to find test project is based off the name of the SDK project/directory. As your test directory does not match.
Your tests will be executed during CI

There is no action needed, but if you are ok with all my comments, let me know and this is ready to be merged

brjohnstmsft commented:
Spoke with @shahabhijeet offline. Since one of the goals of this change is to decouple the development and delivery of these packages, it's better to version them independently and use package references instead of project references. I'll make the changes.

bitsforthought committed:
[Azure Search] Versioning each data plane package separately

After some discussion, we decided it was better for the purposes of decoupled
development and release that we version each of the data plane packages
separately.
